go/runtime/bundle/discovery.go

```diff
@@ -0,0 +1,475 @@
+package bundle
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"maps"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	cmSync "github.com/oasisprotocol/oasis-core/go/common/sync"
+	"github.com/oasisprotocol/oasis-core/go/config"
+)
+
+const (
+	// discoveryInterval is the time interval between (failed) bundle discoveries.
+	discoveryInterval = 15 * time.Minute
+
+	// requestTimeout is the time limit for http client requests.
+	requestTimeout = 10 * time.Second
+
+	// maxBundleSizeBytes is the maximum allowed bundle size in bytes.
+	maxBundleSizeBytes = 20 * 1024 * 1024 // 20 MB
+)
+
+// Discovery is responsible for discovering new bundles.
+type Discovery struct {
+	mu sync.RWMutex
+
+	startOne   cmSync.One
+	discoverCh chan struct{}
+
+	bundleDir      string
+	manifestHashes map[common.Namespace][]hash.Hash
+
+	globalBaseURLs  []string
+	runtimeBaseURLs map[common.Namespace][]string
+	client          *http.Client
+
+	registry Registry
+
+	logger logging.Logger
+}
+
+// NewDiscovery creates a new bundle discovery.
+func NewDiscovery(dataDir string, registry Registry) *Discovery {
+	logger := logging.GetLogger("runtime/bundle/discovery")
+
+	client := http.Client{
+		Timeout: requestTimeout,
+	}
+
+	return &Discovery{
+		startOne:       cmSync.NewOne(),
+		discoverCh:     make(chan struct{}, 1),
+		bundleDir:      ExplodedPath(dataDir),
+		manifestHashes: make(map[common.Namespace][]hash.Hash),
+		client:         &client,
+		registry:       registry,
+		logger:         *logger,
+	}
+}
+
+// Init sets up bundle discovery using node configuration and adds configured
+// and cached bundles to the registry.
+func (d *Discovery) Init() error {
+	// Consolidate all bundles in one place, which could be useful
+	// if we implement P2P sharing in the future.
+	if err := d.copyBundles(); err != nil {
+		return err
+	}
+
+	// Add copied and cached bundles to the registry.
+	if err := d.Discover(); err != nil {
+		return err
```

Comment:
`Init` only adds configured bundles to the registry. In addition, some of them may be already previously exploded (cached)? If this is the case somewhere nested `WriteExploded` will only verify...

However, it does not guarantee to add all the "cached" bundles, as looking at `Discover` method you only add the `*.orc` bundles to the registry (i.e. the ones that were configured). This is desired as some of the exploded bundles may be stale, i.e. no longer part of the config.  (Automatic removal will be handled as part of #5737).

Do I understand your logic correctly? If so I would suggest removing `cached bundles` reference from the `Init` toplevel docs and before `d.Discover()` to avoid confusion?

Thanks for your reply! :)

Reply:
All bundles are copied to the bundle folder.:
- Bundles from the config are copied in `Init` on line 79.
- Bundles fetched from the web are first verified and then copied to the bundle folder.

So the `Discovery` loads all bundles in the config, all bundles that used to be in the config, and all bundles that were downloaded. The idea is that you always have bundle and the extracted folder in the bundle directory. This way, it will be easy to implement p2p2 sharing as one can just check which orc files are in the bundle directory.

Reply:
I see and thanks.

So what I suggest to change (super minor but to avoid confusion): 
```
// Init sets up bundle discovery using node configuration and adds configured
// and cached bundles to the registry.
```
and 
```go
// Add copied and cached bundles to the registry.
```

to
```
// Init sets up bundle discovery using node configuration and adds configured
// bundles (that are guaranteed to be exploded) to the registry.
```
and 
```go
// Add copied bundles (that are guaranteed to be exploded) to the registry.
```

or smtg along the lines, just to emphasize that we are not also moving stale cached bundles (see issue #5737)

Hope it makes sense(?), and feel free to reword as you see fit...

Reply:
> all bundles that used to be in the config

Correct since at init time the `.orc` file (if configured) was moved to `/data/runtimes/bundles` and exploded. 
Note, the nodes that were running previous software (prior to this PR/current version), may have some bundles exploded, later removed them from config (and thus also bundle), so only the exploded part will stay. I guess with #5737 the idea was to remove those stale "exploded" bundles + the ones whose version is now outdated.

Now if we push for the p2p sharing, I guess we go in the other direction, i.e. we further increase the storage load on the nodes as they store essentially all the historical bundles and their exploded paths (technically we could be removing those once outdated and only store bundles). On the positive side we get a more resilient data availability for our software components.

Thinking aloud, if we want to have some kind of p2p sharing we may want only archive nodes responsible for that?  For the normal nodes I don't think we want to store outdated bundles and exploded components. In any case we may want to adapt #5737 cleanup requirements?

Happy to tackle this myself in the follow-up once we align on the strategy.

cc @kostko

Reply:
There is no sense in storing or sharing historic bundles, so any pruning should get rid of them IMO.

Reply:
Yes, all exploded folders without orc file and all orc files that have historic components should be removed.

Is a bundle historic if it contains RONL 1.0 and ROFL 1.0 and there is a new version of the bundle containing RONL1.1 (which can fetch ROFL 1.0 from the previous bundle)?

Reply:
> Is a bundle historic if it contains RONL 1.0 and ROFL 1.0 and there is a new version of the bundle containing RONL1.1 (which can fetch ROFL 1.0 from the previous bundle)?

Only once RONL 1.1 is the active version.

Also note that ROFL components can only be fetched from detached bundles, not from bundles that contain RONLs. If there is a "system ROFL component" (e.g. one that is included in the non-detached bundle) then the expectation is that the new version will include it as well. If it does not, that means that the component has been removed, so in your case the 1.0 version can be pruned.

Reply:
OK, this make sense. RONL + system components in one bundle, and ROFL components in detached bundles.

> Also note that ROFL components can only be fetched from detached bundles, not from bundles that contain RONLs.

Our code doesn't distinguish between ROFL and system components. But it would be nice to have a check that ROFLs are always in a detached bundle, and system components in the runtime bundle.

Reply:
Should we add `Kind = "rofl-system"` or  `Kind = "system"` and verify this in the code?

Reply:
Ups, maybe we don't need this and we just mark the component as system ROFL once is read from an undetached bundle.

Reply:
So "system ROFL component" = non-detached ROFL. There is already the detached flag for this, but maybe it is not obvious? Or there could be an additional component attribute (e.g. `system`), which would only be allowed to be set if the bundle is non-detached?